core/ledger/pvtdatastorage/retroactive_hashed_index.go

```diff
@@ -0,0 +1,176 @@
+			ExpectedFormat: previousDataVersion,
+		})
+	if err != nil {
+		return err
+	}
+
+	defer p.Close()
+
+	for _, l := range ledgerIDs {
+		db := p.GetDBHandle(l)
+		if err := constructHashedIndexFor(l, db); err != nil {
+			return err
+		}
+	}
+	return p.SetDataFormat(currentDataVersion)
+}
+
+// constructHashedIndexFor creates the HashedIndex entries for a given ledger.
+// In this function we also piggyback to upgrade any private data key from format V11 to V12
+func constructHashedIndexFor(ledgerID string, db *leveldbhelper.DBHandle) error {
+	startKey, endKey := entireDatakeyRange()
+	itr, err := db.GetIterator(startKey, endKey)
+	if err != nil {
+		return err
+	}
+	defer itr.Release()
+
+	batch := db.NewUpdateBatch()
+
+	initialTime := time.Now()
+	numEntriesCreated := 0
+	logger.Infow("Starting creation of HashedIndex entries retroactively")
+	for itr.Next() {
+		k := itr.Key()
+		v := itr.Value()
+
+		// In pvtdatastore, we used to persist the data of entire transaction as a single KV in V11.
+		// Later, when we introduced BlockToLive, we started storing each collection data as a KV.
+		// Here we leverage this opportunity (when we create HashedIndexKeys retroactively) to convert
+		// the private data in V11 format, if any so in the next available opportunity,
+		// we will remove the code where we have to detect this difference while processing queries on pvtdatastore.
+		v11Fmt, err := v11Format(k)
+		if err != nil {
+			return err
+		}
+
+		if v11Fmt {
+			blockNum, _, err := v11DecodePK(k)
+			if err != nil {
+				return err
+			}
+
+			txPvtData, err := v11DecodeKV(k, v, nil)
+			if err != nil {
+				return err
+			}
+
+			dataEntries := prepareDataEntries(blockNum, []*ledger.TxPvtData{txPvtData})
+			for _, dataEntry := range dataEntries {
+				var dataKey, dataVal []byte
+				dataKey = encodeDataKey(dataEntry.key)
+				if dataVal, err = encodeDataValue(dataEntry.value); err != nil {
+					return err
+				}
+				batch.Put(dataKey, dataVal)
+				if err := addHashedIndexEntriesInto(batch, dataEntry.key, dataEntry.value); err != nil {
+					return err
+				}
+				numEntriesCreated++
+			}
+			batch.Delete(k)
+		} else {
+			dataKey, err := decodeDatakey(k)
+			if err != nil {
+				return err
+			}
+
+			dataVal, err := decodeDataValue(v)
+			if err != nil {
+				return err
+			}
+			if err := addHashedIndexEntriesInto(batch, dataKey, dataVal); err != nil {
+				return err
+			}
+			numEntriesCreated++
+		}
+
+		if batch.Size() >= maxUpgradeBatchSize {
+			if err := db.WriteBatch(batch, true); err != nil {
+				return err
+			}
+			if (time.Since(initialTime) / time.Second) > 5 {
+				initialTime = time.Now()
+				logger.Infow("Creating HashedIndex entries retroactively...", "total entries created", numEntriesCreated)
+			}
+			batch.Reset()
+		}
+	}
+
+	if err := db.WriteBatch(batch, true); err != nil {
+		return err
+	}
+	logger.Infow("Creatiion of HashedIndex entries retroactively done", "total entries created", numEntriesCreated)
```

Comment:
```suggestion
	logger.Infow("Creation of HashedIndex entries retroactively done", "total entries created", numEntriesCreated)
```

After the long discussion I wanted to merge quick but found this typo :-)

Might as well at channel name (ledgerId) to each of the new log messages as well.

Reply:
> After the long discussion I wanted to merge quick but found this typo :-)

Oh no :-(